test-data/unit/check-isinstance.test

```diff
@@ -2573,3 +2573,36 @@ if issubclass(x, B):
 else:
     reveal_type(x)        # N: Revealed type is "Type[__main__.A]"
 [builtins fixtures/isinstance.pyi]
+
+[case testTypeEqualsCheck]
+from typing import Any
+
+y: Any
+if type(y) == int:
+    reveal_type(y) # N: Revealed type is "builtins.int"
+
+
+[case testMultipleTypeEqualsCheck]
+from typing import Any
+
+x: Any
+y: Any
+if type(x) == type(y) == int:
+    reveal_type(y) # N: Revealed type is "builtins.int"
+    reveal_type(x) # N: Revealed type is "builtins.int"
+
+[case testTypeEqualsCheckUsingIs]
+from typing import Any
+
+y: Any
+if type(y) is int:
+    reveal_type(y) # N: Revealed type is "builtins.int"
+
+[case testTypeEqualsNarrowingUnionWithElse]
+from typing import Union
+
+x: Union[int, str]
+if type(x) is int:
+    reveal_type(x)  # N: Revealed type is "builtins.int"
+else:
+    reveal_type(x)  # N: Revealed type is "Union[builtins.int, builtins.str]"
```

Comment:
Test `type(x) is not int` with `Union[int, str]` (also `!=`). It shouldn't do anything, since the type could be a subclass of `int`.

Optionally, if we are dealing with a `@final` class (`TypeInfo` with `is_final` true), we can actually restrict the type away in the else block. If the type of `x` is `Union[C, D]`, and `C` is final, the type of `x` in the else block could be just `D`, since there can be no subclasses of `C`:

```py
x: Union[C, D]
if type(x) is C:
    reveal_type(x)  # C
else:
    reveal_type(x)  # D, but only if C is final
```

It's fine to leave the handling of final classes as a follow-up issue, though. The current PR already improves things.

Reply:
When you say that `type(x) is not int` shouldn't do anything, do you mean that it shouldn't do any narrowing, like this:
```python
x: Union[int, str]
if type(x) is not int:
    reveal_type(x)  # Union[int, str]
else:
    reveal_type(x)  # Union[int, str]
```
Or do you mean that it should flip what it infers for the if and else cases:
```python
x: Union[int, str]
if type(x) is not int:
    reveal_type(x)  # Union[int, str]
else:
    reveal_type(x)  # int
```

I think it would be the second one because I think the only way for `type(x) is not int` to be false is for x to be an `int`.

Reply:
Yeah, the second is right, I was confused!